### ISSUE TITLE
P38 silently no-op'd on TurboQuant KV path (compile-time capture by aot_compile_fullgraph)

### ISSUE THREAD

**Issue by noonghunna** [Sandermage/genesis-vllm-patches#14]
**P38 silently no-op'd on TurboQuant KV path (compile-time capture)**

We've been investigating a tool-prefill OOM on Qwen3.6-27B AutoRound INT4 + TurboQuant 3-bit KV + MTP K=3 (long-text variant on a single RTX 3090, mem-util 0.975, max-model-len 185K). The OOM site at 50K-token mock tool prefills was at `turboquant_attn.py:903` — `v_full = torch.cat([v_cached_trim.to(qdtype), val_chunk], dim=0)`, exactly the line P38 is designed to eliminate. The dispatcher reports P38 applied at boot:

```
[INFO:genesis.wiring.p38_tq_continuation_memory] [Genesis P38] rebound TurboQuantAttentionImpl._continuation_prefill (persistent K_full/V_full buffers replace torch.cat peak)
[INFO:genesis.apply_all] [Genesis] applied: P38 TQ _continuation_prefill persistent workspace ...
```

To verify whether P38's replacement was actually executing, instrumented `_genesis_continuation_prefill` with a call counter:

```python
global _GENESIS_P38_CALL_COUNT
try:
    _GENESIS_P38_CALL_COUNT += 1
except NameError:
    _GENESIS_P38_CALL_COUNT = 1
if _GENESIS_P38_CALL_COUNT <= 5 or _GENESIS_P38_CALL_COUNT % 100 == 0:
    log.info("[Genesis P38 ACTIVE call=%d] q_len=%d cached_len=%d seq_len=%d",
             _GENESIS_P38_CALL_COUNT, query.shape[0], cached_len, seq_len)
```

Booted long-text and ran a 33K-token tool-prefill stress (which forces multi-chunk continuation prefills — verified passing). **The instrumentation log line never fired**. `_genesis_continuation_prefill` was not called once, despite the rebind succeeding at boot and the original `_continuation_prefill` clearly being invoked (we've seen the OOM trace through it at higher prefill sizes).

Confirmed by inspecting the live patched `turboquant_attn.py` at line 903 inside the running container — it's still the original `torch.cat` line.

**Architectural cause (we suspect):** vLLM's `aot_compile_fullgraph` captures the `forward → _prefill_attention → _continuation_prefill` call chain at compile time, baking in the original method bodies. P38's class-attribute rebind (`TurboQuantAttentionImpl._continuation_prefill = _genesis_continuation_prefill`) updates the live class but the compiled artifact still references the original. This is the same class of issue as PN12 forward_native (which you fixed via the PN25 `torch.library.custom_op` approach on the dev branch — independently converged with our `patch_pn12_compile_safe_custom_op.py` sidecar).

**Why this may not surface on your PROD stack:** your documented PROD configs (35B-A3B-FP8 and 27B-Lorbus-fp8_e5m2) use fp8 KV, not TurboQuant. The entire `TurboQuantAttentionImpl._continuation_prefill` path is inactive on fp8, so P38 reports "applied" but never has a call site to take effect on. We're hitting it because our 27B AutoRound INT4 + TurboQuant 3-bit KV configs (long-text / long-vision / bounded-thinking variants) exercise this exact path.

**Suggested fix:** mirror what PN25 does for `SiluAndMul.forward_native` — register `_continuation_prefill` as a `torch.library.custom_op` so Inductor treats it as opaque and dispatches via the registered op (which CAN be replaced). Our reference implementation for PN12 → PN25 pattern is at [`models/qwen3.6-27b/vllm/patches/patch_pn12_compile_safe_custom_op.py`](https://github.com/noonghunna/club-3090/blob/master/models/qwen3.6-27b/vllm/patches/patch_pn12_compile_safe_custom_op.py) if useful as a reference.

**Practical impact on our shipped configs:** the 33K-token stress class passes anyway (the line 903 cliff is only reached at ~50K-token single-shot prefills, beyond what realistic agent workloads emit). But P38 isn't currently buying us anything on the TQ KV path — the persistent K_full/V_full workspace allocation is reserved (~700 MiB on 27B at 185K context) but never used.

Happy to test a `torch.library.custom_op`-based P38 variant on our rig if you push one to the dev branch — same A/B harness we used for PN25.

Reproducer config: `noonghunna/club-3090@df91d64` master branch, `models/qwen3.6-27b/vllm/compose/docker-compose.long-text.yml` (185K + 0.975 + TQ3 KV + MTP K=3 + GENESIS_ENABLE_P37=1). Stress: `CONTAINER=vllm-qwen36-27b-long-text bash scripts/verify-stress.sh PREFILL_TARGET_CHARS=200000 SKIP_LONGCTX=1`. Full diagnostic in our [docs/CLIFFS.md "Update 2026-05-01 PM — P38 silently no-op'd on TurboQuant KV path"](https://github.com/noonghunna/club-3090/blob/master/docs/CLIFFS.md).

**Comment by Sandermage** [comment]
Hey @noonghunna — this is exactly the right diagnosis. Just landed the fix on `dev`.

**P38B (compile-safe in-source hook)** — Genesis-original 2026-05-01.

Mechanism: text-patch the upstream `turboquant_attn.py` source to inject a delegate hook at the start of `_continuation_prefill` body. The hook calls a class-attached dispatcher (`_genesis_p38_dispatch`) that returns the Genesis impl result OR `None` to fall through to original body.

Key insight from your trace: **source-level text-patch survives `aot_compile_fullgraph` capture** because the compiler reads our modified source at engine init. Class-attribute rebind (P38's mechanism) updates only the live class dict, not the compiled artifact.

Hook structure (injected after the docstring close):
```python
_genesis_p38b_disp = type(self).__dict__.get('_genesis_p38_dispatch', None)
if _genesis_p38b_disp is not None:
    _r = _genesis_p38b_disp(self, layer, query, key_chunk, val_chunk,
                            kv_cache, block_table, cached_len, seq_len,
                            Pi, centroids)
    if _r is not None:
        return _r
# original body resumes here
```

Composition with P38: P38 still rebinds for eager-mode callers (V1 plain forward path); P38B handles compile-mode via the hook. Both share the SAME `_genesis_continuation_prefill` impl. Recommended pairing: enable both.

Status: opt-in via `GENESIS_ENABLE_P38B_COMPILE_SAFE=1`. Default OFF. Idempotent + drift-aware.

Files:
- [`vllm/_genesis/wiring/perf_hotfix/patch_38b_compile_safe_hook.py`](https://github.com/Sandermage/genesis-vllm-patches/blob/dev/vllm/_genesis/wiring/perf_hotfix/patch_38b_compile_safe_hook.py)
- Dispatcher entry `P38B` in [`vllm/_genesis/dispatcher.py`](https://github.com/Sandermage/genesis-vllm-patches/blob/dev/vllm/_genesis/dispatcher.py)

Same `torch.library.custom_op` route you took for PN12 → PN25 was an option — text-patch is simpler here because `_continuation_prefill` has many self-attribute deps and module-level imports that complicate the functional-input contract custom_op needs. Source-edit short-circuit before original body runs reaches the same end (compile-time visibility).

Independent convergence — your `patch_pn12_compile_safe_custom_op.py` via custom_op + Genesis P38B via in-source hook are both valid. Same problem class, two viable mechanisms. Worth documenting both for future operators choosing per case.

Validation pending on our 27B PROD container (boot in progress). Will post bench data.

Closing this ticket once empirical validation passes; meanwhile fix is live on dev branch.

Thanks for the precise diagnosis with stack-frame line numbers — saved a lot of triage time.
